pkg/cluster/kubernetes/upgrade_checks.go

```diff
@@ -0,0 +1,354 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package kubernetes
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/fatih/color"
+	"github.com/gertd/go-pluralize"
+	"github.com/hashicorp/go-multierror"
+	"github.com/siderolabs/gen/slices"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"github.com/siderolabs/talos/pkg/machinery/resources/k8s"
+)
+
+// K8sUpgradeChecks is a set of checks to run before upgrading k8s components.
+type K8sUpgradeChecks struct {
+	talosClient       *client.Client
+	controlPlaneNodes []string
+	log               func(string, ...interface{})
+
+	upgradePath         string
+	upgradeVersionCheck map[string]componentChecks
+}
+
+type k8sUpgradeCheckErr error
+
+type componentChecks struct {
+	// feature gates are common to kube-apiserver, kube-controller-manager and kube-scheduler
+	removedFeatureGates []string
+	// checks specific to kube-apiserver
+	kubeAPIServerChecks apiServerCheck
+	// checks specific to kube-controller-manager
+	kubeControllerManagerChecks k8sComponentCheck
+	// checks specific to kube-scheduler
+	kubeSchedulerChecks k8sComponentCheck
+}
+
+type apiServerCheck struct {
+	// removedAPIResources represent the Kuberenetes API resources that are removed in the upgrade version
+	removedAPIResources []string
```

Comment:
I don't see this being used?

And we had some prior code which did these checks?